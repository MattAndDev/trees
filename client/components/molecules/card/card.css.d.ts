declare namespace CardCssNamespace {
  export interface ICardCss {
    imgFull: string
    imgLoading: string
    loadingEffect: string
    loadingTitle: string
    small: string
    wrap: string
  }
}

declare const CardCssModule: CardCssNamespace.ICardCss & {
  /** WARNING: Only available when `css-loader` is used without `style-loader` or `mini-css-extract-plugin` */
  locals: CardCssNamespace.ICardCss
}

export = CardCssModule
