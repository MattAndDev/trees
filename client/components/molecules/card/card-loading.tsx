import { h, FunctionalComponent, JSX } from 'preact'
import s from './card.css'
import cn from 'classnames'
import { Heading } from '@client/components/atoms'

export const CardLoading: FunctionalComponent<
  JSX.HTMLAttributes<HTMLDivElement>
> = ({ className, ...rest }) => {
  return (
    <div className={cn(s.wrap, className)} {...rest}>
      <Heading className={s.loadingTitle} level={2}>
        &nbsp;
        <span className={cn(s.loadingEffect, s.small)} />
      </Heading>
      <Heading className={s.loadingTitle} level={3}>
        &nbsp;
        <span className={s.loadingEffect} />
      </Heading>
    </div>
  )
}
