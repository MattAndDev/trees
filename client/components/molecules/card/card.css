.wrap {
  padding: 1rem;
  border: 0.1rem solid var(--color-secondary);
  border-radius: 0.5rem;
}

.imgFull {
  width: 100%;
  margin-bottom: 1rem;
}

.imgLoading {
  visibility: hidden;
  position: absolute;
  z-index: -1;
}

.loadingTitle {
  position: relative;
}

.loadingEffect {
  position: absolute;
  top: 0;
  left: 0;
  right: 20%;
  bottom: 0;
  background-color: var(--color-foreground);
  opacity: 0.5;
}

.loadingEffect.small {
  right: 40%;
}
