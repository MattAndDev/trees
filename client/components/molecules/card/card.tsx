import { h, FunctionalComponent, JSX } from 'preact'
import s from './card.css'
import cn from 'classnames'
import { Button, Heading } from '../../atoms'
import { useState } from 'preact/hooks'

export type CardProps = JSX.HTMLAttributes<HTMLDivElement> & {
  title: string
  subTitle: string
  img: string
  btnText?: {
    showImg: string
    hideImg: string
  }
  imgErrorText?: string
}

export const Card: FunctionalComponent<CardProps> = ({
  className,
  title,
  subTitle,
  img,
  btnText = {
    showImg: 'Show image',
    hideImg: 'Hide image'
  },
  ...rest
}) => {
  const [imgVisible, setImgVisible] = useState(false)
  const [imgLoaded, setImgLoaded] = useState(false)
  const [imgLoadError, setImgLoadError] = useState(false)
  return (
    <div className={cn(s.wrap, className, {})} {...rest}>
      <Heading level={2}>{title}</Heading>
      <Heading level={3}>{subTitle}</Heading>
      {imgVisible && (
        <img
          onLoad={() => setImgLoaded(true)}
          className={cn(s.imgFull, { [s.imgLoading]: !imgLoaded })}
          onError={() => setImgLoadError(true)}
          src={img}
        />
      )}
      {!imgLoadError && (
        <Button
          isLoading={imgVisible && !imgLoaded}
          onClick={() => setImgVisible(!imgVisible)}
        >
          {imgVisible && imgLoaded ? btnText.hideImg : btnText.showImg}
        </Button>
      )}
      {imgLoadError && <span>{imgLoadError}</span>}
    </div>
  )
}
