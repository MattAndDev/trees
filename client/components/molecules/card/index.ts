export * from './card'
export * from './card-loading'
