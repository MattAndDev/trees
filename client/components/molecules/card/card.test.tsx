import { h } from 'preact'
import { render, fireEvent, waitFor, screen } from '@testing-library/preact'
import { Card } from './card'

it('Renders card correctly', () => {
  const { container } = render(
    <Card
      title="Foo"
      subTitle="bar"
      img="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg=="
    />
  )
  const elem = container.firstChild as HTMLDivElement
  expect(elem).toMatchSnapshot()
})

it('Toggles image on click', async () => {
  const { container } = render(
    <Card
      title="Foo"
      subTitle="bar"
      btnText={{ showImg: 'show', hideImg: 'hide' }}
      img="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg=="
    />
  )
  const elem = container.firstChild as HTMLDivElement
  fireEvent.click(screen.getByText('show'))
  expect(elem).toMatchSnapshot()
  await waitFor(() => {
    expect(elem.querySelectorAll('img')).toHaveLength(1)
    fireEvent.click(screen.getByText('hide'))
    expect(elem.querySelectorAll('img')).toHaveLength(0)
  })
})
