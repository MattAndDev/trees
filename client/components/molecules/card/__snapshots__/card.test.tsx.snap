// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders card correctly 1`] = `
<div
  class="wrap"
>
  <h2
    class="common h2"
  >
    Foo
  </h2>
  <h3
    class="common h3"
  >
    bar
  </h3>
  <button
    class="wrap"
    type="button"
  >
    Show image
  </button>
</div>
`;

exports[`Toggles image on click 1`] = `
<div
  class="wrap"
>
  <h2
    class="common h2"
  >
    Foo
  </h2>
  <h3
    class="common h3"
  >
    bar
  </h3>
  <img
    class="imgFull imgLoading"
    src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg=="
  />
  <button
    class="wrap"
    type="button"
  >
    show
    <span
      class="spinner"
    />
  </button>
</div>
`;
